File: 2.Strings/C/countVowelsConsonants.c
/**
 * Vowels and Consonants Counter
 * 
 * This program counts the number of vowels and consonants in a given string.
 * It ignores non-alphabetic characters and is case-insensitive.
 * 
 * Algorithm:
 * 1. Iterate through each character of the string
 * 2. Convert the character to lowercase
 * 3. Check if the character is alphabetic
 * 4. If it's a vowel (a, e, i, o, u), increment vowel counter
 * 5. If it's a consonant, increment consonant counter
 * 
 * Time Complexity: O(n) where n is the length of the string
 * Space Complexity: O(1) as we only use two counter variables
 * 
 * Input: A string - "Hello World"
 * Output: Count of vowels and consonants in the string
 */

#include <stdio.h>
#include <ctype.h>  // For tolower() and isalpha() functions

/**
 * countVowelsAndConsonants - Count vowels and consonants in a string
 * @str: The input string to analyze
 * 
 * This function counts and prints the number of vowels and consonants
 * in the given string. It ignores non-alphabetic characters.
 */
void countVowelsAndConsonants(const char *str) {
    int vowels = 0, consonants = 0;
    
    // Process each character in the string
    while (*str) {
        // Convert to lowercase for case-insensitive comparison
        char ch = tolower(*str);
        
        // Only count alphabetic characters
        if (isalpha(ch)) {
            // Check if the character is a vowel
            if (ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u') {
                vowels++;
            } else {
                consonants++;
            }
        }
        // Move to the next character
        str++;
    }
    
    // Display the results
    printf("Vowels: %d\nConsonants: %d\n", vowels, consonants);
}

int main() {
    // Test string
    char str[] = "Hello World";
    
    // Display the input string
    printf("String: %s\n", str);
    
    // Count and display vowels and consonants
    countVowelsAndConsonants(str);
    
    return 0;
}

File: 2.Strings/java/CountVowelsConsonants.java
/**
 * CountVowelsConsonants
 * 
 * This class provides two different methods to count the number of vowels and consonants in a string:
 * 1. Traditional approach using a loop and conditional checks
 * 2. Modern approach using Java Streams and regular expressions
 * 
 * Time Complexity:
 * - Traditional approach: O(n) where n is the length of the string
 * - Streams approach: O(n) where n is the length of the string, but with additional overhead
 *   from stream operations and regex matching
 * 
 * Space Complexity: O(1) for the traditional approach, O(n) for the streams approach due to
 * intermediate stream operations
 * 
 * Input: A string containing letters, numbers, and special characters
 * Output: The count of vowels and consonants in the string
 */
public class CountVowelsConsonants {
    /**
     * Counts the number of vowels and consonants in a string using a traditional approach.
     * 
     * Algorithm:
     * 1. Convert the string to lowercase to handle both upper and lower case letters uniformly
     * 2. Iterate through each character in the string
     * 3. Check if the character is a letter (ignoring numbers, spaces, and special characters)
     * 4. If it's a letter, check if it's a vowel (a, e, i, o, u) or a consonant
     * 5. Increment the appropriate counter
     * 
     * @param str The input string to analyze
     */
    public static void countVowelsAndConsonants(String str) {
        int vowels = 0, consonants = 0;
        // Convert to lowercase for case-insensitive comparison
        str = str.toLowerCase();
        
        // Iterate through each character
        for (int i = 0; i < str.length(); i++) {
            char ch = str.charAt(i);
            // Only count alphabetic characters
            if (Character.isLetter(ch)) {
                // Check if the character is a vowel
                if (ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u') {
                    vowels++;
                } else {
                    // If it's a letter but not a vowel, it's a consonant
                    consonants++;
                }
            }
            // Non-alphabetic characters are ignored
        }
        
        // Display the results
        System.out.println("Vowels: " + vowels);
        System.out.println("Consonants: " + consonants);
    }
    
    /**
     * Counts the number of vowels and consonants in a string using Java Streams and regular expressions.
     * This is a more modern, functional approach but may be less efficient for simple cases.
     * 
     * Algorithm:
     * 1. Define regex patterns for vowels and consonants
     * 2. Convert the string to lowercase
     * 3. Convert the string to a stream of characters
     * 4. Map each character to a string
     * 5. Filter for characters matching the vowel pattern and count them
     * 6. Repeat for consonants
     * 
     * @param str The input string to analyze
     */
    public static void countVowelsAndConsonantsUsingStreams(String str) {
        // Regular expression patterns for vowels and consonants
        String vowelPattern = "[aeiou]";
        String consonantPattern = "[b-df-hj-np-tv-z]"; // All consonants in the English alphabet
        
        // Count vowels using streams
        long vowels = str.toLowerCase().chars()  // Convert to lowercase and get stream of chars
                        .mapToObj(ch -> String.valueOf((char)ch))  // Convert each char to String
                        .filter(ch -> ch.matches(vowelPattern))  // Keep only vowels
                        .count();  // Count the matches
                        
        // Count consonants using streams
        long consonants = str.toLowerCase().chars()
                            .mapToObj(ch -> String.valueOf((char)ch))
                            .filter(ch -> ch.matches(consonantPattern))
                            .count();
        
        // Display the results
        System.out.println("Using Streams:");
        System.out.println("Vowels: " + vowels);
        System.out.println("Consonants: " + consonants);
    }
    
    /**
     * Main method to test the vowel and consonant counting functions with various examples.
     * 
     * @param args Command line arguments (not used)
     */
    public static void main(String[] args) {
        // Test cases with different types of strings
        String[] testStrings = {
            "Hello World",           // Mixed case with space
            "OpenAI GPT",           // Mixed case with uppercase acronym
            "Programming is fun",    // Longer sentence with spaces
            "Java Streams"           // Another example
        };
        
        // Test each string with both methods
        for (String str : testStrings) {
            System.out.println("\nString: " + str);
            System.out.println("Traditional approach:");
            countVowelsAndConsonants(str);
            countVowelsAndConsonantsUsingStreams(str);
        }
    }
}